src/xb-silo.c

```diff
@@ -1530,3 +1538,57 @@ xb_silo_new (void)
 {
 	return g_object_new (XB_TYPE_SILO, NULL);
 }
+
+/**
+ * xb_silo_lookup_query:
+ * @self: an #XbSilo
+ * @xpath: an XPath query string
+ * @error: return location for a #GError, or %NULL
+ *
+ * Create an #XbQuery from the given @xpath XPath string, or return it from the
+ * query cache in the #XbSilo.
+ *
+ * If creating the #XbQuery fails, an error will be propagated from
+ * xb_query_new().
+ *
+ * This function is thread-safe.
+ *
+ * Returns: (transfer full): an #XbQuery representing @xpath
+ * Since: 0.3.0
+ */
+XbQuery *
+xb_silo_lookup_query (XbSilo *self, const gchar *xpath, GError **error)
+{
+	XbSiloPrivate *priv = GET_PRIVATE (self);
+	g_autoptr(GRWLockReaderLocker) locker = g_rw_lock_reader_locker_new (&priv->query_cache_mutex);
+	XbQuery *result;
+
+	result = g_hash_table_lookup (priv->query_cache, xpath);
+	if (result != NULL) {
+		g_object_ref (result);
+		g_debug ("Found cached query ‘%s’ (%p) in silo %p", xpath, result, self);
+	} else {
+		g_autoptr(GRWLockWriterLocker) write_locker = NULL;
+		g_autoptr(XbQuery) query = NULL;
+
+		/* check again with an exclusive lock */
+		g_clear_pointer (&locker, g_rw_lock_reader_locker_free);
+		write_locker = g_rw_lock_writer_locker_new (&priv->query_cache_mutex);
+		result = g_hash_table_lookup (priv->query_cache, xpath);
+		if (result != NULL) {
+			g_object_ref (result);
+			g_debug ("Found cached query ‘%s’ (%p) in silo %p", xpath, result, self);
+		} else {
+			query = xb_query_new (self, xpath, error);
+			if (query == NULL)
+				return NULL;
+			result = g_object_ref (query);
+
+			g_hash_table_insert (priv->query_cache, g_strdup (xpath), g_steal_pointer (&query));
+			g_debug ("Caching query ‘%s’ (%p) in silo %p; query cache now has %u entries",
+				 xpath, query, self, g_hash_table_size (priv->query_cache));
+		}
+	}
+
+	return result;
```

Comment:
would it be better to ref the result here?

Reply:
On the `xb_query_new()` path, the query has to be reffed with the write lock held. Moving the ref to the bottom of the function would mean moving the `GRWLockWriterLocker` to the top scope in the function, which I think would make it less clear where the locked sections need to be.